### PR TITLE
Add dataset from image

### PR DESCRIPTION
**Pull request by Rdornier**
Hello, 

@romainGuiet and I though that having access to dataset information directly from an image could be a good idea.
For example, we get an image from omero, process it on fiji and upload the processed data into the same dataset. For this simple workflow, we need to have the dataset id.

We found a piece of code [here](https://forum.image.sc/t/get-dataset-id-from-image-id-in-script-from-omero/54293) and we implemeted it in the `ImageWrapper` Class. 
As we have the code to get the dataset id, we also think about a method to get the project id from the image id.

Here is a small groovy script to show how it works.
``` javascript
#@String(label="Host", value="host", persist=false) host
#@Integer(label="Port", value = 4064) port
#@String(label="Username", value="demo", persist=false) USERNAME
#@String(label="Password", style='password' , value=PASSWORD , persist=false) PASSWORD
#@Long(label="Image ID", value=119273) id

// create the client and connect to the host
Client user_client = new Client()
user_client.connect(host, port, USERNAME, PASSWORD.toCharArray())

if (user_client.isConnected()){
	println "Connected "+ host
	
	try{
		// Create the image Wrapper object
		ImageWrapper img_wp = user_client.getImage(id)
		
		// Get dataset information from the image ID
		ArrayList<DatasetData> ds = img_wp.getParentDatasets(user_client, id)
		
		// print dataset ID and dataset name
		ds.each{println("dataset name : "+it.getName()+", dataset id : "+it.getId())};
		
	} finally{
		user_client.disconnect()
		println "Disonnected from "+host
	}
	
}


import fr.igred.omero.*
import fr.igred.omero.repository.*
import omero.gateway.model.DatasetData;
```

Do you think that it is possible to integrate this into your project ?
Thank you very much,

Best,
Rémy.

### CODE REVIEW

File: src/main/java/fr/igred/omero/repository/ImageWrapper.java [hunk 1]
```diff
@@ -618,4 +617,31 @@ public List<File> download(Client client, String path)
...
+     *
+     * @param client The client handling the connection.
+     * @param imageId The image ID coming from OMERO
+     *
+     * @return see above
+     *
+     * @throws DSOutOfServiceException
+     * @throws ServerError
+     */
+    public ArrayList<DatasetData> getParentDatasets(Client client, Long imageId) throws DSOutOfServiceException, ServerError {
```
Comment (ppouchin) [inline src/main/java/fr/igred/omero/repository/ImageWrapper.java:632]: Returns `ArrayList<DatasetData>` and throws `DSOutOfServiceException` as well as `ServerError`, which forces downstream to include OMERO Gateway dependencies

File: src/main/java/fr/igred/omero/repository/ImageWrapper.java [hunk 2]
```diff
@@ -618,4 +617,31 @@ public List<File> download(Client client, String path)
...
+        IQueryPrx qs = client.getGateway().getQueryService(client.getCtx());
+        ParametersI p = new ParametersI();
+        p.addLong("id", imageId);
+        List<IObject> objs = qs.findAllByQuery("select l.parent from DatasetImageLink as l " +
```
Comment (ppouchin) [inline src/main/java/fr/igred/omero/repository/ImageWrapper.java:636]: Can be replaced by `Client#findByQuery`

File: simple-omero-client.iml [hunk 3]
```diff
@@ -0,0 +1,121 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ppouchin) [inline simple-omero-client.iml:1]: Not sure this file should be added.

**Review by ppouchin — CHANGES_REQUESTED** [review]
I'll work on the library tomorrow, as I still have issues to close.

You're right that this feature should have been integrated, but the way I see it, something like the following code should be enough:

```java



    /**
     * Retrieves the datasets containing this image
     *
     * @param client The client handling the connection.
     *
     * @return See above.
     *
     * @throws OMEROServerError   Server error.
     * @throws ServiceException   Cannot connect to OMERO.
     * @throws AccessException    Cannot access data.
     * @throws ExecutionException A Facility can't be retrieved or instantiated.
     */
    public List<DatasetWrapper> getDatasets(Client client)
    throws OMEROServerError, ServiceException, AccessException, ExecutionException {
        List<IObject> os = client.findByQuery("select link.parent from DatasetImageLink as link " +
                                              "where link.child=" + getId());
        return client.getDatasets(os.stream().map(IObject::getId).map(RLong::getValue).toArray(Long[]::new));
    }
```

Although tests are required too, for automated testing. 

I'll see if I can edit this PR or integrate the code after merging the other branch tomorrow.

**Comment by Rdornier** [comment]
Thank you very much for your answer ; the code is cleaner now and works perfectly !

>Although tests are required too (for CI).

Here is a test for the new method. However, I'm not able to run the test on my computer/server because all the parameters are hard-coded for another server for which I do not have access. I hope it will run for you. If you have any suggestions on how I can run tests on my server, feel free to tell me.

```Java
    @Test
    public void testGetDatasetsFromImages() throws Exception {
        assertEquals(DATASET1.id, client.getImage(IMAGE1.id).getDatasets(client).get(0).getId());
    }
```

**Comment by ppouchin** [comment]
> If you have any suggestions on how I can run tests on my server, feel free to tell me.

The tests actually use [omero-test-infra](https://github.com/ome/omero-test-infra), which relies on Docker. 

To run the tests, you need to first clone this repo to your working directory:
```shell
git clone https://github.com/openmicroscopy/omero-test-infra .omero
```

Then, I advise you to set a DOCKER_ARGS variable to mount the `target` and `~/.m2` directories as folders in Docker, e.g. (on Linux/UNIX):
```shell
export DOCKER_ARGS="-v $HOME/.m2:/root/.m2 -v $(pwd)/target:/src/target -t"
```

Finally, to run build and tests, you just need to put this command:
```shell
.omero/lib-docker
```

These steps are what is done in GitHub actions, as seen in the [config file](https://github.com/GReD-Clermont/simple-omero-client/blob/main/.github/workflows/maven.yml).

On my computer, once I have cloned the omero-test-infra repo, I only use the following script from the project repository:
```shell
#!/bin/sh

export DOCKER_ARGS="-v $HOME/.m2:/root/.m2 -v $(pwd)/target:/src/target -t"
.omero/lib-docker
```
(If I'm on Windows, I do this using WSL.)


I'm thinking of switching to plain "docker-compose", to be able to run tests easily from my IDE, but omero-test-infra eases the population process.

**Comment by Rdornier** [comment]
Hello, 

Ok. Do I have to do something for the file deletion ?

I run the build file to perform the tests on the Github CI. I had to modify the expected number of projects from 2 to 3 because it failed to pass the `testGetAllProjects() `in `ClientTest` class. Now, all tests have passed but I get the following issue and I don't know how to manage it. It happens during "Sonar analysis".

```
Error:  Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.7.0.1746:sonar (default-cli) on project simple-omero-client: Project not found. Please check the ‘sonar.projectKey’ and ‘sonar.organization’ properties, the ‘SONAR_TOKEN’ environment variable, or contact the project administrator -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
```

**Comment by ppouchin** [comment]
> Ok. Do I have to do something for the file deletion ?

This deletion may impact your local copy as the file will get deleted when you pull the modifications. If you recreate the file, do not add it to the repo if possible.
I prefer not to have IDE-specific files in the repo, but maybe I should at least push a `.editorconfig` (which may have IDE-specific options though). 

> I had to modify the expected number of projects from 2 to 3 because it failed to pass the `testGetAllProjects() `in `ClientTest` class.

This was because I had forgotten a "delete" in one of the tests I added recently. It was fixed in the latest commit on the main branch.

> Now, all tests have passed but I get the following issue and I don't know how to manage it. It happens during "Sonar analysis".

Ok. That comes from the fact that this project uses [SonarCloud](https://sonarcloud.io/project/branches_list?id=GReD-Clermont_simple-omero-client) to monitor code quality & code security. The GitHub actions expects an "action secret" to sync with the project on SonarCloud. I'll see if I can make it optional somehow, by checking if the secret is set or not.

**Comment by ppouchin** [comment]
I have made adjustments to your PR and will make a commit, if it's ok for you. 
I will try to quickly adjust the GitHub actions (for SonarCloud and CodeCov) before pushing.

**Comment by Rdornier** [comment]
Ok thanks for the answer.
I add in the .gitignore file to not commit IDE files (*.iml).  Sorry for that, it's the first time I use git and github and I'm not perfectly fluent with all things.
Ok for the commit. Thanks.

**Review by ppouchin — APPROVED** [review]

**Comment by ppouchin** [comment]
Ok. This has been merged, thank you for your contribution. 
I still need to solve an issue, and then, I will release version 5.9.1.

**Pull request merged**